go/store/prolly/sort/external.go

```diff
@@ -0,0 +1,458 @@
+// limitations under the License.
+
+package sort
+
+import (
+	"bufio"
+	"container/heap"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"sync"
+
+	"github.com/dolthub/dolt/go/store/util/tempfiles"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+const maxRetries = 2
+
+// tupleSorter inputs a series of unsorted tuples and outputs a sorted list
+// of tuples. Batches of tuples sorted in memory are written to disk, and
+// then k-way merge sorted to produce a final sorted list. The |fileMax|
+// parameter limits the number of files spilled to disk at any given time.
+// The maximum memory used will be |fileMax| * |batchSize|.
+type tupleSorter struct {
+	keyCmp    func(val.Tuple, val.Tuple) bool
+	files     [][]keyIterable
+	inProg    *keyMem
+	fileMax   int
+	fileCnt   int
+	batchSize int
+	tmpProv   tempfiles.TempFileProvider
+}
+
+func NewTupleSorter(batchSize, fileMax int, keyCmp func(val.Tuple, val.Tuple) bool, tmpProv tempfiles.TempFileProvider) *tupleSorter {
+	if fileMax%2 == 1 {
+		// round down to even
+		// fileMax/2 will be compact parallelism
+		fileMax -= 1
+	}
+	ret := &tupleSorter{
+		fileMax:   fileMax,
+		batchSize: batchSize,
+		keyCmp:    keyCmp,
+		tmpProv:   tmpProv,
+	}
+	ret.inProg = newKeyMem(batchSize)
+	return ret
+}
+
+func (a *tupleSorter) Flush(ctx context.Context) (iter keyIterable, err error) {
+	// don't flush in-progress, just sort in memory
+	a.inProg.sort(a.keyCmp)
+
+	if len(a.files) == 0 {
+		// don't go to disk if we didn't reach a mem flush
+		return a.inProg, nil
+	}
+
+	var iterables []keyIterable
+	iterables = append(iterables, a.inProg)
+	for _, level := range a.files {
+		for _, file := range level {
+			iterables = append(iterables, file)
+		}
+	}
+
+	newF, err := a.newFile()
+	if err != nil {
+		return nil, err
+	}
+	allKeys := newKeyFile(newF, a.inProg.byteLim)
+	m, _ := newFileMerger(ctx, a.keyCmp, allKeys, iterables...)
+	if err := m.run(ctx); err != nil {
+		return nil, err
+	}
+	return allKeys, nil
+}
+
+func (a *tupleSorter) Insert(ctx context.Context, k val.Tuple) (err error) {
+	if !a.inProg.insert(k) {
+		if err := a.flushMem(ctx); err != nil {
+			return err
+		}
+		a.inProg.insert(k)
+	}
+	return
+}
+
+func (a *tupleSorter) flushMem(ctx context.Context) error {
+	// flush and replace |inProg|
+	if a.inProg.Len() > 0 {
+		newF, err := a.newFile()
+		if err != nil {
+			return err
+		}
+		newFile, err := a.inProg.flush(newF, a.keyCmp)
+		if err != nil {
+			return err
+		}
+		a.inProg = newKeyMem(a.batchSize)
+		if len(a.files) == 0 {
+			a.files = append(a.files, []keyIterable{newFile})
+		} else {
+			a.files[0] = append(a.files[0], newFile)
+		}
+		a.fileCnt++
+	}
+	for level, ok := a.shouldCompact(); ok; level, ok = a.shouldCompact() {
+		if err := a.compact(ctx, level); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *tupleSorter) newFile() (*os.File, error) {
+	f, err := a.tmpProv.NewFile("", "key_file_")
```

Comment:
Something seems off about the lifetime of these files. We put them in keyIterables, and those return the Iter, but the iter's Close() method doesn't close the file. And there's no way provided in the API to close and clean them up in the case of an error. And it seems like we would want to os.Remove() these files after we're done with them, since we had a contract was to use a certain number of files?

Near at hand is something like:

1) Make it explicit that keyFile.IterAll() can only ever be called once.
2) On keyFile.IterAll(), set keyFile's `f *os.File` to `nil`.
3) `func (k *keyFile) Close() { if k.f != nil { k.f.Close(); os.Remove(k.f.Name()) }`
4) keyFileReader gets the `*os.File` as well as its buffered reader.
5) `func (r keyFileReader) Close() { r.f.Close(); os.Remove(r.f.Name()) }`
6) `func (a *tupleSorter) Close() { for _, fs := range a.files { for _, f := range fs { f.Close() } } }`

Or something like that?